integration-tests/src/test_loop/tests/syncing.rs

```diff
@@ -0,0 +1,154 @@
+// to sync up to the current state of the network.
+#[test]
+fn test_sync_from_genesis() {
+    init_test_logger();
+    let builder = TestLoopBuilder::new();
+
+    let initial_balance = 10000 * ONE_NEAR;
+    let accounts =
+        (0..100).map(|i| format!("account{}", i).parse().unwrap()).collect::<Vec<AccountId>>();
+    let clients = accounts.iter().take(NUM_CLIENTS).cloned().collect_vec();
+
+    let mut genesis_builder = TestGenesisBuilder::new();
+    genesis_builder
+        .genesis_time_from_clock(&builder.clock())
+        .protocol_version_latest()
+        .genesis_height(10000)
+        .gas_prices_free()
+        .gas_limit_one_petagas()
+        .shard_layout_simple_v1(&["account3", "account5", "account7"])
+        .transaction_validity_period(1000)
+        .epoch_length(10)
+        .validators_desired_roles(&clients.iter().map(|t| t.as_str()).collect_vec(), &[])
+        .shuffle_shard_assignment_for_chunk_producers(true);
+    for account in &accounts {
+        genesis_builder.add_user_account_simple(account.clone(), initial_balance);
+    }
+    let genesis = genesis_builder.build();
+
+    let TestLoopEnv { mut test_loop, datas: node_datas, tempdir } =
+        builder.genesis(genesis.clone()).clients(clients).build();
+
+    let first_epoch_tracked_shards = {
+        let clients = node_datas
+            .iter()
+            .map(|data| &test_loop.data.get(&data.client_sender.actor_handle()).client)
+            .collect_vec();
+        clients.tracked_shards_for_each_client()
+    };
+    tracing::info!("First epoch tracked shards: {:?}", first_epoch_tracked_shards);
+
+    execute_money_transfers(&mut test_loop, &node_datas, &accounts);
+
+    // Make sure the chain progresses for several epochs.
+    let client_handle = node_datas[0].client_sender.actor_handle();
+    test_loop.run_until(
+        |test_loop_data| {
+            test_loop_data.get(&client_handle).client.chain.head().unwrap().height > 10050
+        },
+        Duration::seconds(50),
+    );
+
+    // Make a new TestLoopEnv, adding a new node to the network, and check that it can properly sync.
+    let mut stores = Vec::new();
+    for data in &node_datas {
+        stores.push((
+            test_loop
+                .data
+                .get(&data.client_sender.actor_handle())
+                .client
+                .chain
+                .chain_store
+                .store()
+                .clone(),
+            None,
+        ));
+    }
+    stores.push((create_test_store(), None)); // new node starts empty.
+
+    // Properly shut down the previous TestLoopEnv.
+    // We must preserve the tempdir, since state dumps are stored there,
+    // and are necessary for state sync to work on the new node.
+    TestLoopEnv {
+        test_loop,
+        datas: node_datas,
+        tempdir: TempDir::new().unwrap(), /* don't destroy yet */
+    }
+    .shutdown_and_drain_remaining_events(Duration::seconds(20));
```

Comment:
Consider copying code from `shutdown_and_drain_remaining_events`:
```
for node_data in self.datas {
    self.test_loop.data.get_mut(&node_data.state_sync_dumper_handle).stop();
}

self.test_loop.shutdown_and_drain_remaining_events(timeout);
```

Fake tempdir looks more fragile.

Or make `shutdown_and_drain_remaining_events` return that tempdir, letting caller to decide whether to drop it or not.